vi/day20.md

```diff
@@ -0,0 +1,484 @@
+
+```ts
+/**
+ * output: 
+ * - next: MouseEvent {...}
+ */
+fromEvent(btn, 'click').subscribe(observer);
+
+/**
+ * output: 
+ * - next: MouseEvent {...}
+ */
+fromEventPattern(
+  (handler) => {
+    btn.addEventListener('click', handler);
+  },
+  (handler) => {
+    btn.removeEventListener('click', handler);
+  }
+).subscribe(observer);
+```
+
+Another example:
+
+```ts
+/**
+ * output:
+ * - next: 10 10
+ */
+fromEvent(btn, 'click')
+  .pipe(map((ev: MouseEvent) => ev.offsetX + ' ' + ev.offsetY))
+  .subscribe(observer);
+
+// fromEventPattern
+// In this example, we'll break fromEventPattern() arguments into functions.
+
+function addHandler(handler) {
+  btn.addEventListener('click', handler);
+}
+
+function removeHandler(handler) {
+  btn.removeEventListener('click', handler);
+}
+
+/**
+ * output:
+ * - next: 10 10
+ */
+fromEventPattern(
+  addHandler,
+  removeHandler,
+  (ev: MouseEvent) => ev.offsetX + ' ' + ev.offsetY
+).subscribe(observer);
+```
+
+From these examples, we can tell that `fromEventPattern()` accepts 3 arguments: `addHandler`, `removeHandler`, and an optional `projectFunction`. Moreover, the behavior isn't that much different from `fromEvent()`.
+
+`fromEventPattern()` gives you the ability to convert the Events using their _original_ APIs, like we're using the DOM API like `addEventListener` and `removeEventListener` to convert the `MouseEvent` to `Observable`. With this knowledge, you can apply `fromEventPattern()` with more complex Event APIs like `SignalR Hub`.
+
+```ts
+// _getHub() is an arbitrary function that returns a Hub
+const hub = this._getHub(url);
+
+return fromEventPattern(
+  (handler) => {
+    // open the Websocket
+    hub.connection.on(methodName, handler);
+
+    if (hub.refCount === 0) {
+      hub.connection.start();
+    }
+
+    hub.refCount++;
+  },
+  (handler) => {
+    hub.refCount--;
+    // close the Websocket on Unsubscribe
+    hub.connection.off(methodName, handler);
+    if (hub.refCount === 0) {
+      hub.connection.stop();
+    }
+  }
+);
+```
+
+or `SocketIO`
+
+```ts
+export interface SocketEventMap {
+  event1: string;
+  event2: number;
+}
+
+@Injectable({providedIn: 'root'})
+export class SocketService {
+  private socket: Socket;
+
+  // init logic
+
+  on<EventName extends keyof SocketEventMap>(event: EventName): Observabe<SocketEventMap<EventName>> {
+    return fromEventPattern(
+      (handler) => {
+        this.socket?.on(event, handler);
+      },
+      (handler) => {
+        this.socket?.off(event, handler);
+      }
+    );
+  }
+}
+
+// usage
+this.socketService.on('event1'); // Observable<string>
+this.socketService.on('event2'); // Observable<number>
+```
+
+## `fromFetch()`
```

Comment:
Should we mention how `fromFetch` internally differs from ordinary `fetch`? It uses an `AbortController` to tear down when the subscriber unsubscribes https://rxjs.dev/api/fetch/fromFetch

Reply:
Nice catch. It'd also be worth mention that this allows for Cancellation since `fromFetch()` is lazy.